### PR TITLE
closes #319

### LINKED ISSUE #319: error reading unicode symbols in the vcf.gz file
Hello @daler!

Thanks for developing and maintaining pybedtools!

We discovered a bug when running pybedtools in bcbio: https://github.com/bcbio/bcbio-nextgen/issues/3078

test.py:
```
import pybedtools

vrn_file = "small.vcf.gz"
spliceslop = "S10SJ.out-minimized-padded.bed"
bt = pybedtools.BedTool(vrn_file).intersect(spliceslop, wa=True, header=True, v=True).saveas("out.bed")
```
bed and vcf files are here: 
https://drive.google.com/open?id=1p1EuXxU2X7280dzKScohW_JWF55B16Xn

error message:
```
Traceback (most recent call last):
  File "test-error.py", line 5, in <module>
    bt = pybedtools.BedTool(vrn_file).intersect(spliceslop, wa=True, header=True, v=True).saveas("out.bed")
  File "/n/app/bcbio/dev/anaconda/lib/python3.6/site-packages/pybedtools/bedtool.py", line 840, in decorated
    result = method(self, *args, **kwargs)
  File "/n/app/bcbio/dev/anaconda/lib/python3.6/site-packages/pybedtools/bedtool.py", line 3134, in saveas
    compressed=compressed)
  File "/n/app/bcbio/dev/anaconda/lib/python3.6/site-packages/pybedtools/bedtool.py", line 1289, in _collapse
    out_.writelines(in_)
  File "/n/app/bcbio/dev/anaconda/lib/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 6389: invalid continuation byte
```

we think this clinvar record triggers the issue:
`clinvar_disease_name=Ch�diak-Higashi_syndrome`

What do you think would be the best approach to tackle this issue?
maybe adding `errors=ignore` to 
https://github.com/daler/pybedtools/blob/master/pybedtools/bedtool.py#L1406
as recommended here: 
https://stackoverflow.com/questions/1883604/reading-utf-8-characters-from-a-gzip-file-in-python/19794943#19794943

Thanks!
Sergey

### PR DESCRIPTION
**Pull request by naumenko-sa**

### CODE REVIEW

**Comment by daler** [comment]
Thanks! The failing tests are due to py27 (e.g. https://travis-ci.org/daler/pybedtools/jobs/647596751#L1480). I plan to deprecate py27 support in the next release, but that will take some effort to remove all the py27 bits. In the meantime, I think the solution is to get py27 working here.

**Comment by naumenko-sa** [comment]
Seems to be working now? @daler could you please merge?

File: pybedtools/bedtool.py [hunk 1]
```diff
@@ -1403,10 +1403,17 @@ def _collapse(
...
-                        out_.write(trackline.strip() + "\n")
+                in_ = in_open_func(iterable.fn, "rt")
+                if trackline:
+                    out_.write(trackline.strip() + "\n")
+                try:
                     out_.writelines(in_)
+                except UnicodeDecodeError:
+                    in_.close()
+                    with in_open_func(iterable.fn, "rt", errors="ignore") as in_:
+                        out_.writelines(in_)
```
Comment (daler) [inline pybedtools/bedtool.py]: If the error isn't seen until e.g. line 100 out of 1000, won't this cause the output file have 100 + 1000 lines?

File: pybedtools/test/data/issue319.out.bed [hunk 2]
```diff
@@ -0,0 +1,3496 @@
...
...20/atanasova/work/variation/rnaseq/gatk-haplotype/S7-gatk-haplotype.vcf.gz)]), (RodBindingCollection [(RodBinding name=variant72 source=/n/scratch2/hsph_bioinformatic_core/sn240/atanasova2020/atanasova/work/variation/rnaseq/gatk-haplotype/S8-gatk-haplotype.vcf.gz)]), (RodBindingCollection [(RodBinding name=variant73 source=/n/scratch2/hsph_bioinformatic_core/sn240/atanasova2020/atanasova/work/variation/rnaseq/gatk-haplotype/S9-gatk-haplotype.vcf.gz)])] out=/n/scratch2/hsph_bioinformatic_core/sn240/atanasova2020/atanasova/work/bcbiotx/tmp_y699b1y/b1-chr1_0_248956422.vcf.gz includeNonVariantSites=false uniquifySamples=false annotateNDA=false useNewAFCalculator=false heterozygosity=0.001 indel_heterozygosity=1.25E-4 heterozygosity_stdev=0.01 standard_min_confidence_threshold_for_calling=10.0 standard_min_confidence_threshold_for_emitting=30.0 max_alternate_alleles=4 max_genotype_count=1024 max_num_PL_values=100 input_prior=[] sample_ploidy=2 annotation=[] group=[StandardAnnotation] dbsnp=(RodBinding name=dbsnp source=/n/app/bcbio/dev/genomes/Hsapiens/hg38/variation/dbsnp-151.vcf.gz) filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false">
```
Comment (daler) [inline pybedtools/test/data/issue319.out.bed]: This line should be taken out -- these test files are distributed with the pybedtools package.

Are all the headers of this file required for reproducing the test? I'd prefer to only have what's needed for a minimally-reproducible example.

File: pybedtools/test/data/issue319.out.bed [hunk 3]
```diff
@@ -0,0 +1,3496 @@
...
+##contig=<ID=chrUn_KN707636v1_decoy,length=1725>
+##contig=<ID=chrUn_KN707637v1_decoy,length=5354>
+##contig=<ID=chrUn_KN707638v1_decoy,length=2189>
+##contig=<ID=chrUn_KN707639v1_decoy,length=1294>
+##contig=<ID=chrUn_KN707640v1_decoy,length=1831>
+##contig=<ID=chrUn_KN707641v1_decoy,length=1647>
+##contig=<ID=chrUn_KN707642v1_decoy,length=2943>
+##contig=<ID=chrUn_KN707643v1_decoy,length=2857>
+##contig=<ID=chrUn_KN707644v1_decoy,length=1030>
+##contig=<ID=chrUn_KN707645v1_decoy,
```
Comment (daler) [inline pybedtools/test/data/issue319.out.bed]: In the rendered text here,`Chdiak-Higashi_syndrome` does not appear to be unicode (it's missing the `é`). Is this due to github rendering? Or is the file missing the the problematic unicode (in which case a passing test is misleading?)

Comment (naumenko-sa) [reply pybedtools/test/data/issue319.out.bed]: Hi Ryan @daler !

Thanks for you suggestions. Sorry for the delay. Fixing this is still a priority for us in bcbio.

I've simplified the error handling and removed unnecessary lines from the test vcf header.
The test works - without a fix it breaks pybedtools, with the fix it passes.

The problematic symbol is rendered as 
problematic_symbol
in the terminal or in Atom.

So it is in the file - it breaks pybedtools when given as input.

Please merge.
Sergey

**Comment by daler** [comment]
Great, thanks.

Are you sure you want [this line](https://github.com/naumenko-sa/pybedtools/blob/master/pybedtools/test/data/issue319.out.bed#L41) and [this other line](https://github.com/naumenko-sa/pybedtools/blob/master/pybedtools/test/data/issue319.out.bed#L3409)? They contain path information that will make it into every future installation of pybedtools.

**Comment by naumenko-sa** [comment]
Thanks, I removed these two as well. It works without them.

**Comment by daler** [comment]
Thanks!

**Comment by naumenko-sa** [comment]
Thanks @daler! 
Could you cut a new release at some point?

**Pull request merged**